Libraries/LibWeb/Streams/AbstractOperations.cpp

```diff
@@ -50,6 +65,359 @@ GC::Ref<SizeAlgorithm> extract_size_algorithm(JS::VM& vm, QueuingStrategy const&
+void set_up_cross_realm_transform_readable(JS::Realm& realm, ReadableStream& stream, HTML::MessagePort& port)
+{
+    // 1. Perform ! InitializeReadableStream(stream).
+    initialize_readable_stream(stream);
+
+    // 2. Let controller be a new ReadableStreamDefaultController.
+    auto controller = realm.create<ReadableStreamDefaultController>(realm);
+
+    // 3. Add a handler for port’s message event with the following steps:
+    add_message_event_listener(realm, port, HTML::EventNames::message,
+        [&port, controller](JS::VM& vm, HTML::MessageEvent const& message) {
+            // 1. Let data be the data of the message.
+            auto data = message.data();
+
+            // 2. Assert: data is an Object.
+            VERIFY(data.is_object());
+
+            // 3. Let type be ! Get(data, "type").
+            auto type = MUST(data.get(vm, vm.names.type));
+
+            // 4. Let value be ! Get(data, "value").
+            auto value = MUST(data.get(vm, vm.names.value));
+
+            // 5. Assert: type is a String.
+            auto type_string = type.as_string().utf8_string_view();
+
+            // 6. If type is "chunk",
+            if (type_string == "chunk"sv) {
+                // 1. Perform ! ReadableStreamDefaultControllerEnqueue(controller, value).
+                MUST(readable_stream_default_controller_enqueue(controller, value));
+            }
+            // 7. Otherwise, if type is "close",
+            else if (type_string == "close"sv) {
+                // 1. Perform ! ReadableStreamDefaultControllerClose(controller).
+                readable_stream_default_controller_close(controller);
+
+                // 2. Disentangle port.
+                port.disentangle();
+            }
+            // 8. Otherwise, if type is "error",
+            else if (type_string == "error"sv) {
+                // 1. Perform ! ReadableStreamDefaultControllerError(controller, value).
+                readable_stream_default_controller_error(controller, value);
+
+                // 2. Disentangle port.
+                port.disentangle();
+            }
+        });
+
+    // 4. Add a handler for port’s messageerror event with the following steps:
+    add_message_event_listener(realm, port, HTML::EventNames::messageerror,
+        [&realm, &port, controller](JS::VM&, HTML::MessageEvent const&) {
+            // 1. Let error be a new "DataCloneError" DOMException.
+            auto error = WebIDL::DataCloneError::create(realm, "Unable to transfer stream"_string);
+
+            // 2. Perform ! CrossRealmTransformSendError(port, error).
+            cross_realm_transform_send_error(realm, port, error);
+
+            // 3. Perform ! ReadableStreamDefaultControllerError(controller, error).
+            readable_stream_default_controller_error(controller, error);
+
+            // 4. Disentangle port.
+            port.disentangle();
+        });
+
+    // FIXME: 5. Enable port’s port message queue.
```

Comment:
I believe this is basically just the same thing as calling `start()` on the message queue, which is basically a no-op with our current implementation (but I believe we need to change that at some point). When I had implemented the steps I just made a method on `MessagePort` that just called `start()` for this.

Reply:
I think it's the other way around - `start()` calls the enable steps.

The enable steps do a bunch of event loop / task shenanigans so I figured I'd kick that can down the road for now.
https://html.spec.whatwg.org/multipage/web-messaging.html#port-message-queue

Reply:
Right, I had just made a no-op function thing for it, but yeah fair enough to to kick the can. Only reason I mentioned it is since I knew we have one or maybe two places like this: https://github.com/LadybirdBrowser/ladybird/blob/7af188dc52e06ca3a22705ccd4efd70f1f19bdea/Services/WebWorker/WorkerHost.cpp#L211